flow/connectors/mongo/mongo.go

```diff
@@ -0,0 +1,391 @@
+		return errors.New("MongoDB is not configured as a replica set, which is required for CDC")
+	}
+	if myState, ok := result["myState"].(int32); !ok || myState != 1 {
+		return fmt.Errorf("MongoDB is not the primary node in the replica set, current state: %d", myState)
+	}
+
+	return nil
+}
+
+func (c *MongoConnector) GetTableSchema(
+	ctx context.Context,
+	_ map[string]string,
+	_ protos.TypeSystem,
+	tableMappings []*protos.TableMapping,
+) (map[string]*protos.TableSchema, error) {
+	result := make(map[string]*protos.TableSchema, len(tableMappings))
+	idFieldDescription := &protos.FieldDescription{
+		Name:         "_id",
+		Type:         string(types.QValueKindString),
+		TypeModifier: -1,
+		Nullable:     false,
+	}
+	dataFieldDescription := &protos.FieldDescription{
+		Name:         "_full_document",
+		Type:         string(types.QValueKindJSON),
+		TypeModifier: -1,
+		Nullable:     false,
+	}
+
+	for _, tm := range tableMappings {
+		result[tm.SourceTableIdentifier] = &protos.TableSchema{
+			TableIdentifier:       tm.SourceTableIdentifier,
+			PrimaryKeyColumns:     []string{"_id"},
+			IsReplicaIdentityFull: true,
+			System:                protos.TypeSystem_Q,
+			NullableEnabled:       false,
+			Columns: []*protos.FieldDescription{
+				idFieldDescription,
+				dataFieldDescription,
+			},
+		}
+	}
+
+	return result, nil
+}
+
+// stubs for CDCPullConnectorCore
+func (c *MongoConnector) EnsurePullability(ctx context.Context, req *protos.EnsurePullabilityBatchInput) (
+	*protos.EnsurePullabilityBatchOutput, error,
+) {
+	return nil, nil
+}
+
+func (c *MongoConnector) ExportTxSnapshot(context.Context, map[string]string) (*protos.ExportTxSnapshotOutput, any, error) {
+	return nil, nil, nil
+}
+
+func (c *MongoConnector) FinishExport(any) error {
+	return nil
+}
+
+func (c *MongoConnector) SetupReplication(ctx context.Context, input *protos.SetupReplicationInput) (model.SetupReplicationResult, error) {
+	// TODO: need to think of the situation where the database is idle and we don't "get" a resume token.
+	changeStreamOpts := options.ChangeStream().
+		SetComment("PeerDB changeStream").
+		SetFullDocument(options.UpdateLookup).
+		SetFullDocumentBeforeChange(options.WhenAvailable)
+	changeStream, err := c.client.Watch(ctx, mongo.Pipeline{}, changeStreamOpts)
+	if err != nil {
+		return model.SetupReplicationResult{}, fmt.Errorf("failed to start change stream for storing initial resume token: %w", err)
+	}
+	defer changeStream.Close(ctx)
+	var resumeToken bson.Raw
+	changeStream.ResumeToken()
+	for changeStream.Next(ctx) {
+		resumeToken = changeStream.ResumeToken()
+		if resumeToken != nil {
+			c.SetLastOffset(ctx, input.FlowJobName, model.CdcCheckpoint{
+				Text: base64.StdEncoding.EncodeToString(resumeToken),
+			})
+		}
+	}
+	if err := changeStream.Err(); err != nil {
+		if errors.Is(err, context.DeadlineExceeded) {
+			c.logger.Info("SetupReplication context deadline exceeded, stopping change stream")
```

Comment:
this logs "stopping change stream", but function does not early returns, same with the "else if" branch below.